File: MyLeetCodeAnswers/C++/Permutations.cpp

/*看onenote笔记，优化之后的代码*/
class Solution {
public:
    vector<vector<int> > result;
    vector<vector<int> > permute(vector<int> &num) {
        result.clear();
        permute(num, 0);
        return result;
    }
    
    void permute(vector<int> &num, int start) {
        if (num.size() == start+1) {
            // find one!
            result.push_back(num);
            return ;
        }
        for (int i = start; i < num.size(); ++i) {
            // swap
            swap(num[start], num[i]);
            //recursive
            permute(num, start+1);
            //swap back
            swap(num[start], num[i]);
        }
    }
};

/*直写，比较丑，没有使用系统swap方法，递归调用多此一举*/
class Solution {
public:
    vector<vector<int> > permute(vector<int> &num) {
        return permute(num, 0);
    }
    
    vector<vector<int> > permute(vector<int> &num, int start) {
        vector<vector<int> > result;
        if (num.size() <= 0) return result;
        if (num.size() == start+1) {
            result.push_back(vector<int>(1, num[start]));
            return result;
        }
        for (int i = start; i < num.size(); ++i) {
            // swap
            int t = num[start];
            num[start] = num[i];
            num[i] = t;
            //recursive
            vector<vector<int> > tmp = permute(num, start+1);
            for (int k=0; k < tmp.size(); ++k) {
                tmp[k].push_back(num[start]);
            }
            result.insert(result.end(), tmp.begin(), tmp.end());
            //swap back
            t = num[start];
            num[start] = num[i];
            num[i] = t;
        }
    }
};

File: MyLeetCodeAnswers/java/Permutations.java
import java.util.ArrayList;
import java.util.List;

/**
 * Source   : https://leetcode.com/problems/permutations/
 * Author   : Wang Meng
 * Date     : 2016-03-20
 * Given a collection of distinct numbers, return all possible permutations.
 * <p>
 * For example,
 * [1,2,3] have the following permutations:
 * [1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], and [3,2,1].
 */
public class Permutations {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ret = new ArrayList<>();
        if (nums == null || nums.length <= 0) return ret;
        permute(nums, 0, ret);
        return ret;
    }

    private void permute(int[] nums, int start, List<List<Integer>> ret) {
        if (start == nums.length - 1) {
            List<Integer> one = new ArrayList<>();
            for (int num : nums) {
                one.add(num);
            }
            ret.add(one);
            return;
        }
        for (int i = start; i < nums.length; ++i) {
            swap(nums, start, i);
            permute(nums, start + 1, ret);
            swap(nums, start, i);
        }
    }

    private void swap(int[] nums, int a, int b) {
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
}
